### ISSUE TITLE
jsx-key support array methods

### ISSUE THREAD

**Issue by giladgray** [palantir/tslint-react#115]
`jsx-key` does not flag this element as missing `key`:
```tsx
const myArray: JSX.Element[] = [];
myArray.push(
    <li className="missing-key">no tslint error</li>
);
return <ul>{myArray}</ul>
```

**Comment by adidahiya** [comment]
this is doable, but requires making the rule implementation an "optionally typed rule" which uses the type checker when it's available

**Comment by adidahiya** [comment]
Closing due to age and deprecation timeline, see #210
